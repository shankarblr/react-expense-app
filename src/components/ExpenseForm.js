import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'


export default class ExpenseForm extends React.Component{
    constructor(props) {
        super(props)
        this.state ={
            description : props.expense ? props.expense.description : '',
            amount : props.expense ? props.expense.amount.toString() :'',
            note:props.expense ? props.expense.note :'',
            createdAt : props.expense ? moment(props.expense.createdAt) :moment(),
            calenderfocused :false,
            error:'',
            isEdit :props.expense?true:false 
        }
    }
    
    onDescriptionChange =(e) => {
        const desc = e.target.value;
        this.setState(() => (
            { 
                description :desc 
            }));
    };

    onAmountChange =(e) => {
        const amt = e.target.value
        
        if(!amt || amt.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ( {   amount : amt   }))
        }
    };

    onNoteChange =(e) => {
        const note = e.target.value
        this.setState(() => (
            {
                note
            }
        ))
        
    }
    onDateChange =(date) =>{
        if(date){
            this.setState(() => ( { createdAt : date }) )
        }
    };

    onFocusChange =({focused}) => {
        this.setState(() => ( { calenderfocused : focused }) )

    }

    onSubmit =(e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.note){
            return this.setState(() => ( { error : 'Please provied description and notes fields' }) )
        }else{
            this.setState(() => ( { error : '' }) )

            this.props.onSubmit({
                description : this.state.description ,
                amount : parseFloat(this.state.amount,10),
                note: this.state.note,
                createdAt : this.state.createdAt.valueOf()
            })

        }

    }

    render() {
        return (
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>} 
                    <input type="text" 
                        placeholder ="description" 
                        autoFocus 
                        className="text-input"
                        value = {this.state.description}
                        onChange ={this.onDescriptionChange} >
                    </input>
                    <input 
                        type="text" 
                        className="text-input"
                        placeholder="amount" 
                        value ={this.state.amount}
                        onChange ={this.onAmountChange}
                        ></input>
                    <SingleDatePicker
                        date = {this.state.createdAt}
                        focused = {this.state.calenderfocused}
                        onFocusChange ={this.onFocusChange}
                        onDateChange ={this.onDateChange}
                        numberOfMonths ={1}
                        isOutsideRange ={() => false}
                    />    
                    <textarea 
                        placeholder="Add a note to expense (optional)"
                        value = {this.state.note}
                        onChange = {this.onNoteChange}
                        className="textarea"
                    >
                    </textarea>
                    <div>
                        <button className="button">{this.state.isEdit? 'Edit Expense' : 'Add Expense' }</button>
                    </div>
                </form>
        )
    };

}