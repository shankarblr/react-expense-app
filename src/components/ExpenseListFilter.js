import React from 'react'
import {connect} from 'react-redux'
import {setTextFilter,setStartDate,setEndDate,sortByDate,sortByAmount} from './../actions/filters'
import { DateRangePicker  } from "react-dates";


class ExpenseListFilter extends React.Component{
    constructor(props){
        super(props)
        this.state = {
            calenderfocused :null
        }
    }
    onDatesChange =({startDate,endDate}) =>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    };
    onFocusChange =(focusedInput) =>{
        this.setState(() => ( { calenderfocused :focusedInput }) )
    };
    onTextFilter = (e) =>{
        this.props.setTextFilter(e.target.value)    
    }
    onSortBy = (e) => {
        e.target.value ==='date' ? this.props.sortByDate() : this.props.sortByAmount()
    }
    render() {
        return (
            <div>
                <input type="text" value ={this.props.filter.text} onChange ={this.onTextFilter}></input>
                <select value ={this.props.filter.sortBy}  onChange = {this.onSortBy} >
                    <option value = "date">Date</option>
                    <option value ="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate = {this.props.filter.startDate}
                    endDate = {this.props.filter.endDate}
                    onDatesChange ={this.onDatesChange}
                    focusedInput={this.state.calenderfocused} 
                    onFocusChange ={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange = {() => false}
                    showClearDates={true}
                />
            </div>
        )
    };
}

const mapStateToExpenseListFilter = (state) => {
    return {
        filter : state.filter
    } 
};

const mapDispatchToProps =(dispatch) => {
    return {
        setTextFilter : (text) => dispatch(setTextFilter(text)),
        sortByDate : () => dispatch(sortByDate()),
        sortByAmount :() => dispatch(sortByAmount()),
        setStartDate : (startDate) => dispatch(setStartDate(startDate)),
        setEndDate : (endDate) => dispatch(setEndDate(endDate))
    };
};

export default connect(mapStateToExpenseListFilter,mapDispatchToProps)(ExpenseListFilter)