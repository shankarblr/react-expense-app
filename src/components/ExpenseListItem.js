import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

export const ExpenseItem = ({dispatch,id,description,amount,createdAt}) => (
            <Link  className="list-item"  to={`/edit/${id}`} >
                <div>
                    <h3 className="list-item__title">{description}</h3>
                    <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
                </div>
                <h3 className="list-item__data">${amount}</h3>
            </Link>
            
);

export default ExpenseItem