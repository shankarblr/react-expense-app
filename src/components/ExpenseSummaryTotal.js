import React from 'react'
import {connect} from 'react-redux'
import getFilterdExpenses from '../selectors/FilterdExpenses';
import selectExpenseTotal from '../selectors/selectExpenseTotal';

const ExpenseSummary = (props) => (
        <div>
            <h1>Viewing {props.totalEpenses} totalling ${props.totalAmount}</h1>
        </div>
);

const mapStateToProps = (state,props) =>{
    const expenses = getFilterdExpenses(state.expenses,state.filter);
    return {
        totalEpenses : expenses.length,
        totalAmount : selectExpenseTotal(expenses)
    };
};
export default connect(mapStateToProps)(ExpenseSummary)

