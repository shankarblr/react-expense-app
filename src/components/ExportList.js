import React from 'react'
import {connect} from 'react-redux'
import ExpenseItem from './ExpenseListItem'
import FilterdExpenses from './../selectors/FilterdExpenses'
import getFilterdExpenses from './../selectors/FilterdExpenses';

export const ExportList = (props) => (
        <div>
            <h1>Export List </h1>
            {  props.expenses.map((e) =>  <ExpenseItem key={e.id} {...e}/>)        }
        </div>
);

const mapStateToProps = (state) => {
    return {
        expenses : getFilterdExpenses(state.expenses,state.filter)
    };
};


export default connect(mapStateToProps)(ExportList)