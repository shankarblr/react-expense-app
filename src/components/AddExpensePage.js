import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import {addExpenses} from './../actions/expenses';

export class AddExpensePage extends React.Component{
    onSubmit = (expense) => {
        this.props.onAddExpense(expense)
        this.props.history.push("/")
    };
    render(){
        return(
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit = {this.onSubmit} />
            </div>
        )
    };
};

const mapDispatchToProps = (dispatch) => ({
    onAddExpense : (expense) => dispatch(addExpenses(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage)