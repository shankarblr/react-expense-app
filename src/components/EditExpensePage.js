import React from 'react'
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {editExpenses} from './../actions/expenses'
import {removeExpenses} from './../actions/expenses'

export class EditExpensePage extends React.Component{
  onSubmit = (expense) => {
    this.props.editExpenses(this.props.expense.id,expense) 
    this.props.history.push("/")
  };
  onRemove = () => {
    this.props.removeExpenses(this.props.expense.id)
    this.props.history.push("/")
  };
  render(){
      return(
            <div>
              <ExpenseForm expense = {this.props.expense} onSubmit = {this.onSubmit} />
              <button onClick ={this.onRemove}>Remove</button>
            </div>
      )
  };
};

const mapStateToEditExpensePage = (state,props) => {
  return {
    expense : state.expenses.find((e) =>  e.id === props.match.params.id  )
  }
}

const mapDispatchToProps =(dispatch,props) => {
  return {
    editExpenses : (id,expense) => dispatch(editExpenses(id,{...expense})),
    removeExpenses : (id) => dispatch(removeExpenses(id ))
  }
};

export default connect(mapStateToEditExpensePage,mapDispatchToProps)(EditExpensePage)