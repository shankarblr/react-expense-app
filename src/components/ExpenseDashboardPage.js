import React from 'react'
import ExportList  from './ExportList'
import ExpenseListFilter from './ExpenseListFilter'

const ExpenseDashboardPage = () =>(
    <div>
        <ExpenseListFilter />
        <ExportList />
    </div>
)

export default ExpenseDashboardPage