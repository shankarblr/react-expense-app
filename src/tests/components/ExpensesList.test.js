import React from 'react'
import { shallow } from 'enzyme'
import {ExportList} from '../../components/ExportList'
import testData from '../fixtures/expensesTestData'

test('should render expenses with list',() =>{
    const wrapper =  shallow(<ExportList expenses = {testData}/>)
    expect(wrapper).toMatchSnapshot();
})


