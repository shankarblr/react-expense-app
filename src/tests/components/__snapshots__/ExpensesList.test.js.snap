// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses with list 1`] = `
<div>
  <h1>
    Export List 
  </h1>
  <ExpenseItem
    amount={100}
    createdAt={0}
    description="Rent"
    id="1"
    key="1"
    note=""
  />
  <ExpenseItem
    amount={200}
    createdAt={-345600000}
    description="Gas"
    id="2"
    key="2"
    note=""
  />
  <ExpenseItem
    amount={400}
    createdAt={691200000}
    description="Food"
    id="3"
    key="3"
    note=""
  />
</div>
`;
